src/crimson/osd/pg.cc

```diff
@@ -500,7 +502,15 @@ seastar::future<> PG::submit_transaction(ObjectContextRef&& obc,
 		      pg_log_entry_t::MODIFY : pg_log_entry_t::DELETE,
 		    obc->obs.oi.soid, osd_op_p.at_version, obc->obs.oi.version,
 		    osd_op_p.user_modify ? osd_op_p.at_version.version : 0,
-		    osd_op_p.req->get_reqid(), osd_op_p.req->get_mtime(), 0);
+		    osd_op_p.req->get_reqid(), osd_op_p.req->get_mtime(),
+                    op_info.allows_returnvec() && !ops.empty() ? ops.back().rval.code : 0);
```

Comment:
how about moving this statement out, and returning the retval as a `future<int>`? so we don't need to repeat exactly the same code to calculate the retval of an op in `PG::do_osd_ops()`? something like
```c++
int result = 0;
if (op_info.allows_returnvec() && !ops.empty()) {
  result = ops.back();
}
log_entries.emplace_back( ///...
                      result,
                      /**/);
peering_state.append_log_with_trim_to_updated(/**/);
return backend->mutate_object(/**/).then(
  [this, result, /**/](/**/) {
  // ...
  return seastar::make_ready_future<int>(result);
}
```

Reply:
Would you be fine with addressing that with a dedicated commit?
I think a bigger refactor is needed and the `TODO` marker is already there:

```cpp
// TODO: refactor the submit_transaction
```